### ISSUE TITLE
Questions regarding UMI processing

### ISSUE THREAD

**Issue by rebber** [fulcrumgenomics/fgbio#434]
Hi,

My team has just started using UMIs for sensitive somatic variant calling from ctDNA sequencing, and I'm therefore now exploring the fgbio tools for processing the UMIs. I have some questions regarding this process.

1. As I understand it GroupReadsByUmi creates groups based on position and UMI. We are concerned about misalignment in repetitive regions and multiple UMI families being created from the same fragment due to some reads not mapping in the correct position. Have you accounted for this in any way in any of the UMI tools? We are thinking about adding a step of local re-alignment (e.g. with GATK IndelRealigner) after mapping with bwa and before GroupReadsByUmi. Any thoughts on that?

2. What was your rationale when deciding to apply the aggressive filtering of reads in GroupReadsByUmi? The stdout when running GroupReadsByUmi says "Filtered out XXX reads that were not part of a high confidence FR mapped read pair.". But in the source code (src/main/scala/com/fulcrumgenomics/umi/GroupReadsByUmi.scala) it seems like there's no filtering on read orientation, only mapping status, mapq and that the mates should be on the same chromosome. Is actually RF, RR & FF reads kept? Or are they discarded somewhere else? The reason I'm asking is because we want to perform structural variant calling as well, and there the those reads will be of highest importance. Do you have any feeling for how the end results will be affected if we were to remove these filters, to keep also reads that aren't of as high quality?

3. After consensus read creation, mapping and filtering we want to finish with clipping the overlapping parts of each read pair. Ultimately we would like to create a consensus of the overlapping ends, or at least mask any non-agreeing bases. Is this possible to achieve with ClipBam? Or do you have any plans to introduce this feature?

I would be grateful for any of your input on these questions!
Best regards,
Rebecka

**Comment by tfenne** [comment]
Hi @rebber.  The answers to your (1) and (2) are related.  

The reason that `GroupReadsByUmi` applies the filtering that it does is to try and avoid the grouping (and thus creation of consensus reads) of reads from a single original molecule into two or more separate groups.  To do this we filter out reads that are mapped cross-chromosomally (we don't feel that consensus reads add value for calling large SVs where this might be useful), and reads where the two ends are not mapped at a reasonable mapq (default mapq>=30) to the same chromosome.  

If you want to access large scale repetitive regions such as gene duplications you really have a couple of choices: lower the mapq cutoff, which is a little risky, or doing something assay-dependent to determine which reads map to which repeat copy and refining mapping and mapping quality based off that.  Or a third option would be to mask out all but one copy during alignment, forcing all reads to one copy of the repeat, and either during re-alignment or genotyping do something to account for the copies/higher ploidy.

`GroupReadsByUmi` only uses the placement of the 5' ends of each read, and will attempt to account for any soft-clipping at the 5' end.  We haven't explored indel realignment to see if that help, but I would expect it's utility to be limited.  Still, it'd be an easy experiment to run.

Re: FR orientation.  I believe you are correct and the log statement is a hangover from an earlier version.  I also don't see anything that would explicitly remove RF/FF/RR read pairs.  I'll update the log message, thanks for catching that.

Re (3), `ClipBam` will only do the clipping.  We talk about what you're suggesting from time to time, and always come back to the conclusion that while it's nice in theory, it won't actually help in practice.  Our experience has always been that a) it's better to filter out low quality bases on the way into consensus calling so that the consensus error rate (the rate of error between raw reads for a consensus and the called consensus) is low, b) you tend to want a minimum of 2-4 reads per consensus to reliably squash out both sequencing and sample-prep errors, and c) that once you have the above there's really nothing to be gained by creating a consensus from overlapping pairs.  That said, I think if you could demonstrate the utility is >> 0 in a real dataset, we'd be happy to add that to our toolset.

**Comment by rebber** [comment]
Thank you for your input!

I will play around some with local realignment, mapq cut-off  etc, and see if there's any gain from it. 
Regarding overlap clipping I think we'll just go for the ClipBam functionality then.

Also thanks for clarification about the non-FR reads!

**Comment by fa-vio** [comment]
Hello, 

Not sure if it is appropriate to continue this thread, or rather open a new issue, but the problem seems to be related.

Working on UMIs , I tried to roughly follow the analysis pipeline outlined [here](http://nilshomer.com/2017/07/05/single-strand-umi-somatic-variant-calling/). After some steps I end up with a mapped and annotated BAM file, where the UMI pattern is annotated with the RX tag. 

Tried to run `GroupReadsByUmi` and obtained an empty result file, with the log message saying all reads were filtered out because they were not part of a high confidence mapped read pair, even though the reads I am working on have a good mapping quality, surely above 30.
I tried with different grouping strategies with no success; tried tuning mapping quality flag, edits and umi-length flags, but nothing changed: the resulting BAM file was empty.

I realized the BQ tag was missing from the annotated BAM, so I run `samtools calmd -r` to add that tag and validated the BAM using Picard's  `ValidateSamFile`, which reports no error. Here a sample read from the BAM file (quite a long row, sorry):

```
    M05223:41:000000000-D46GK:1:1101:18741:24506    99      chr1    115252176       60      151M    =       115252186       151     AAGCTGTACCATACCTGTCTGGTCTTGGCTGAGGTTTCAATGAATGGAATCCCGTAACTCTTGGCCAGTTCGTGGGCTTGTTTTGTATCAACTGTCCTTGTTGGCAAATCACACTTGTTTCCCACTAGCACCATAGGTACATCATCCGAGT AAAABFFFFFFFGGGGGGGGGGHHHHHHHHGGHGGHHHHHHHHHHHGHHHHHHGHGGGIHHHHHHHHHHGHGGGGGGHHGHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHGHHHHHHHHHFHHHHHHHHHHHHHHHHHHHGGGG MD:Z:151        RG:Z:1  NM:i:0  BQ:Z:CC@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@D    AS:i:151        XS:i:0  RX:Z:ATAGGTACCG
```

The UMI is a 10 random bp, and it has been trimmed off from 5' read 2 before creating the BAM file. 

I am writing here because I do not understand why the algorithm is filtering out all the reads, thus I cannot figure a way to solve my problem. Am I missing some steps?
